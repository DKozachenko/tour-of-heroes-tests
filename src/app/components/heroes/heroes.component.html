<h2 automation-id="heading" pw-automation-id="heading">My Heroes</h2>

<div automation-id="create-div" pw-automation-id="create-div">
  <label
    for="new-hero"
    automation-id="create-label"
    pw-automation-id="create-label"
    >Hero name:
  </label>
  <input
    id="new-hero"
    #heroName
    automation-id="create-input"
    pw-automation-id="create-input"
  />

  <!-- (click) passes input value to add() and then clears the input -->
  <button
    type="button"
    class="add-button"
    (click)="add(heroName.value); heroName.value = ''"
    automation-id="add-button"
    pw-automation-id="add-button"
  >
    Add hero
  </button>
</div>

<ul class="heroes" automation-id="heroes" pw-automation-id="heroes">
  <li
    *ngFor="let hero of heroes"
    automation-id="hero-item"
    pw-automation-id="hero-item"
    cy-automation-id="hero-item"
  >
    <a
      routerLink="/detail/{{ hero.id }}"
      automation-id="hero-link"
      pw-automation-id="hero-link"
    >
      <span class="badge">{{ hero.id }}</span> {{ hero.name }}
    </a>
    <button
      type="button"
      class="delete"
      title="delete hero"
      (click)="delete(hero)"
      automation-id="hero-button"
      pw-automation-id="hero-button"
    >
      x
    </button>
  </li>
</ul>
