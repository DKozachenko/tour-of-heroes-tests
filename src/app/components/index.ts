export { DashboardComponent } from './dashboard/dashboard.component';
export { HeroDetailComponent } from './hero-detail/hero-detail.component';
export { HeroSearchComponent } from './hero-search/hero-search.component';
export { HeroesComponent } from './heroes/heroes.component';
export { MessagesComponent } from './messages/messages.component';
export { AppComponent } from './app/app.component';
