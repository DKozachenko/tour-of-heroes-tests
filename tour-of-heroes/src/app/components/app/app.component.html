<h1 automation-id="heading">Tour of Heroes</h1>
<nav automation-id="navigation">
  <a routerLink="/dashboard" automation-id="navigation-link">Dashboard</a>
  <a routerLink="/heroes" automation-id="navigation-link">Heroes</a>
</nav>
<router-outlet></router-outlet>
<app-messages automation-id="messages"></app-messages>
