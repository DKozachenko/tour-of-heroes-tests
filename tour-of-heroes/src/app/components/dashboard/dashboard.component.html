<h2 automation-id="heading" pw-automation-id="heading">Top Heroes</h2>
<div
  class="heroes-menu"
  automation-id="heroes-menu"
  pw-automation-id="heroes-menu"
>
  <a
    *ngFor="let hero of heroes"
    routerLink="/detail/{{ hero.id }}"
    automation-id="hero-link"
    pw-automation-id="hero-link"
  >
    {{ hero.name }}
  </a>
</div>

<app-hero-search
  automation-id="hero-search"
  pw-automation-id="hero-search"
></app-hero-search>
