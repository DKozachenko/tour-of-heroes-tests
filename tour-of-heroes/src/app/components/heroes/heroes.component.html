<h2 automation-id="heading">My Heroes</h2>

<div automation-id="create-div">
  <label for="new-hero" automation-id="create-label">Hero name: </label>
  <input id="new-hero" #heroName automation-id="create-input" />

  <!-- (click) passes input value to add() and then clears the input -->
  <button
    type="button"
    class="add-button"
    (click)="add(heroName.value); heroName.value = ''"
    automation-id="add-button"
  >
    Add hero
  </button>
</div>

<ul class="heroes" automation-id="heroes">
  <li *ngFor="let hero of heroes" automation-id="hero-item">
    <a routerLink="/detail/{{ hero.id }}" automation-id="hero-link">
      <span class="badge">{{ hero.id }}</span> {{ hero.name }}
    </a>
    <button
      type="button"
      class="delete"
      title="delete hero"
      (click)="delete(hero)"
      automation-id="hero-button"
    >
      x
    </button>
  </li>
</ul>
