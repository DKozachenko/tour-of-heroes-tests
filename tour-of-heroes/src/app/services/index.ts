export { HeroService } from './hero/hero.service';
export {
  InMemoryDataService,
  HEROES,
} from './in-memory-data/in-memory-data.service';
export { MessageService } from './message/message.service';
