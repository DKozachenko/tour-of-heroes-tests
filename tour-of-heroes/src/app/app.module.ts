import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemoryDataService } from './services';

import {
  DashboardComponent,
  HeroDetailComponent,
  HeroesComponent,
  HeroSearchComponent,
  MessagesComponent,
} from './components';

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    AppRoutingModule,
    HttpClientModule,

    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
    // and returns simulated server responses.
    // Remove it when a real server is ready to receive requests.
    HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, {
      dataEncapsulation: false,
    }),
  ],
  declarations: [
    AppComponent,
    DashboardComponent,
    HeroesComponent,
    HeroDetailComponent,
    MessagesComponent,
    HeroSearchComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
