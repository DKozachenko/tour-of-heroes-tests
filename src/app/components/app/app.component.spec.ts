import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { fakeAsync, tick } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { Location } from '@angular/common';
import { Router, RouterLink, RouterModule } from '@angular/router';
import {
  MockBuilder,
  MockRender,
  MockedComponentFixture,
  NG_MOCKS_ROOT_PROVIDERS,
  ngMocks,
} from 'ng-mocks';
import { AppComponent } from './app.component';
import { AppModule } from '../../app.module';
import { DashboardComponent } from '../dashboard/dashboard.component';
import { HeroesComponent } from '../heroes/heroes.component';

class PageObject {
  private fixtureDebugElement: DebugElement;

  private getElementByAutomationId(id: string): DebugElement {
    return this.fixtureDebugElement.query(By.css(`[automation-id=${id}]`));
  }

  private getElementsByAutomationId(id: string): DebugElement[] {
    return this.fixtureDebugElement.queryAll(By.css(`[automation-id=${id}]`));
  }

  constructor(fixture: MockedComponentFixture<AppComponent>) {
    this.fixtureDebugElement = fixture.debugElement;
  }

  get heading(): DebugElement {
    return this.getElementByAutomationId('heading');
  }

  get navigation(): DebugElement {
    return this.getElementByAutomationId('navigation');
  }

  get links(): DebugElement[] {
    return this.getElementsByAutomationId('navigation-link');
  }

  get messages(): DebugElement {
    return this.getElementByAutomationId('messages');
  }
}

const LINKS = [
  {
    text: 'Dashboard',
    url: '/dashboard',
  },
  {
    text: 'Heroes',
    url: '/heroes',
  },
];

describe('AppComponent', () => {
  beforeEach(() => {
    return MockBuilder(AppComponent, AppModule);
  });

  function createFixture(): MockedComponentFixture<AppComponent> {
    return MockRender(AppComponent);
  }

  describe('Layout', () => {
    it('should contain heading with "Tour of Heroes" title', () => {
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.heading).not.toBeNull();
      expect(pageObject.heading.nativeElement.textContent).toBe(
        'Tour of Heroes'
      );
      expect(pageObject.heading.nativeElement).toMatchSnapshot();
    });

    it('should contain navigation with links', () => {
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.navigation).not.toBeNull();
      expect(pageObject.links).toHaveLength(LINKS.length);
      expect(pageObject.navigation.nativeElement).toMatchSnapshot();

      for (let i = 0; i < pageObject.links.length; ++i) {
        const navigationLink = pageObject.links[i];
        const link = LINKS[i];
        expect(navigationLink.nativeElement.textContent).toBe(link.text);
        expect(navigationLink.injector.get(RouterLink).routerLink).toBe(
          link.url
        );
        expect(navigationLink.nativeElement).toMatchSnapshot();
      }
    });

    it('should contain "app-messages" component', () => {
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.messages).not.toBeNull();
      expect(pageObject.messages.nativeElement).toMatchSnapshot();
    });
  });
});

describe('AppComponent:Routing', () => {
  beforeEach(() => {
    return MockBuilder(
      [
        AppComponent,
        RouterModule,
        RouterTestingModule.withRoutes([]),
        NG_MOCKS_ROOT_PROVIDERS,
      ],
      AppModule
    );
  });

  function createFixture(): MockedComponentFixture<AppComponent> {
    return MockRender(AppComponent);
  }

  it('should change location if appropriate hero link has clicked', fakeAsync(() => {
    const fixture = createFixture();
    const router = fixture.point.injector.get(Router);
    const location = fixture.point.injector.get(Location);

    if (fixture.ngZone) {
      fixture.ngZone.run(() => router.initialNavigation());
      tick();
    }

    expect(location.path()).toBe(LINKS[0].url);
    expect(() => ngMocks.find(DashboardComponent)).not.toThrow();
    expect(() => ngMocks.find(HeroesComponent)).toThrow();

    const pageObject = new PageObject(fixture);
    for (let i = 0; i < pageObject.links.length; ++i) {
      const navigationLink = pageObject.links[i];

      if (fixture.ngZone) {
        fixture.ngZone.run(() => {
          navigationLink.triggerEventHandler('click', {
            button: 0,
          });
        });
        tick();
      }

      expect(location.path()).toBe(
        navigationLink.injector.get(RouterLink).href
      );
      if (i === 0) {
        expect(() => ngMocks.find(DashboardComponent)).not.toThrow();
      } else if (i === 1) {
        expect(() => ngMocks.find(HeroesComponent)).not.toThrow();
      }
    }
  }));
});
