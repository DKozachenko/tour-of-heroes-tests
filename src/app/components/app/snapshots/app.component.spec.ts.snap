// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent Layout should contain "app-messages" component 1`] = `
<app-messages
  automation-id="messages"
  cy-automation-id="messages-component"
  pw-automation-id="messages-component"
/>
`;

exports[`AppComponent Layout should contain heading with "Tour of Heroes" title 1`] = `
<h1
  automation-id="heading"
>
  Tour of Heroes
</h1>
`;

exports[`AppComponent Layout should contain navigation with links 1`] = `
<nav
  automation-id="navigation"
>
  <a
    automation-id="navigation-link"
    cy-automation-id="dashboard-navigation-link"
    pw-automation-id="dashboard-navigation-link"
    routerlink="/dashboard"
  >
    Dashboard
  </a>
  
  <a
    automation-id="navigation-link"
    cy-automation-id="heroes-navigation-link"
    pw-automation-id="heroes-navigation-link"
    routerlink="/heroes"
  >
    Heroes
  </a>
  
</nav>
`;

exports[`AppComponent Layout should contain navigation with links 2`] = `
<a
  automation-id="navigation-link"
  cy-automation-id="dashboard-navigation-link"
  pw-automation-id="dashboard-navigation-link"
  routerlink="/dashboard"
>
  Dashboard
</a>
`;

exports[`AppComponent Layout should contain navigation with links 3`] = `
<a
  automation-id="navigation-link"
  cy-automation-id="heroes-navigation-link"
  pw-automation-id="heroes-navigation-link"
  routerlink="/heroes"
>
  Heroes
</a>
`;
